Add ploidy summary functions #416

timothymillar opened the pull request:
See #416 

Ac ouple of notes:
- I don't know what I'm missing to populate the "comment" attribute in the HTML REPR?
- I think 'infer' is a better verb than 'impute' in this case.

jeromekelleher commented on sgkit/stats/aggregation.py:
```diff
@@ -546,3 +546,129 @@ def sample_stats(
+    merge: bool = True,
+) -> Dataset:
+    """Infer the ploidy of each genotype call.
+
```
It would be nice to give a little bit more detail here (one sentence) as to how it actually does this

jeromekelleher commented on sgkit/tests/test_aggregation.py:
```diff
@@ -289,3 +292,64 @@ def test_sample_stats(precompute_variant_allele_count):
+    )
+    # test as fixed ploidy
+    cp = infer_call_ploidy(ds).call_ploidy
+    np.testing.assert_equal(cp, np.array([[4, 4, 4], [4, 4, 4]]))
```
Just as a passing thing: I don't think you need to explicitly convert the second array to numpy, 
```python
  np.testing.assert_equal(cp, [[4, 4, 4], [4, 4, 4]])
```
also works.

timothymillar replied:
It's actually throwing `ValueError: The truth value of an array ...` when used with `xr.DataArray` and `list`. `cp.values` would work but calling `np.array` is consistent with the other tests.

jeromekelleher replied:
Ah, I didn't try with an xr array. My bad!

jeromekelleher reviewed (approved):
LGTM, just needs a little bit more explanation on the docs side.

tomwhite commented:
> * I don't know what I'm missing to populate the "comment" attribute in the HTML REPR?

Try using `create_dataset` from `sgkit.utils` rather than calling the `Dataset` constructor directly. This was added in #420

timothymillar commented:
Thanks @jeromekelleher and @tomwhite 
I have added helper method `infer_non_alleles` to make things a little more consistent between mixed and fixed ploidy datasets.
This can result in nested use of `define_variable_if_absent` to define both `call_genotype_non_allele` and `call_ploidy` which might not be ideal but i think it's the most consistent approach.

jeromekelleher reviewed (approved):
LGTM

tomwhite commented:
@jeromekelleher is this ready to be merged?

jeromekelleher commented:
> @jeromekelleher is this ready to be merged?

Yep, sorry, this one fell off my stack. I've added the label
